src/App.js

```diff
@@ -37,77 +128,429 @@ function App(props) {
         }
     }
 
-    // Counter used for unique list item ids
-    // Increment every time a list item is added
-    const [counter, setCounter] = useState(500);
+    // Hidden modals for adding/editing/deleting list items
+    const [showAddItemModal, setShowAddItemModal] = useState(false);
+    const [showDeleteItemModal, setShowDeleteItemModal] = useState(false);
+
+    // Hidden modals for adding/editing/deleting lists
+    const [showAddListModal, setShowAddListModal] = useState(false);
+    const [showEditListModal, setShowEditListModal] = useState(false);
+    const [showDeleteListModal, setShowDeleteListModal] = useState(false);
 
-    const [showModal, setShowModal] = useState(false);
+    // Add item
+    const onAddItem = (text, priority) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.Item,
+            op: UndoOp.Add,
+            oldList: subcollectionName,
+            newList: null,
+            data: itemsData,
+        }]));
 
-    // Change between showing non-completed items vs. all items
-    const onShowToggle = () => {
-        setShowState(NextShowState[showState]);
+        const id = generateUniqueID();
+        setDoc(doc(db, collectionPath + subcollectionName, id), {
+            id: id,
+            text: text,
+            completed: false,
+            priority: priority,
+            created: serverTimestamp(),
+        });
+    }
+
+    // Edit item
+    const onEditItem = (id, text, completed, priority) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.Item,
+            op: UndoOp.Edit,
+            oldList: subcollectionName,
+            newList: null,
+            data: itemsData
+        }]));
+
+        updateDoc(doc(db, collectionPath + subcollectionName, id), {
+           text: text,
+           completed: completed,
+           priority: priority,
+        });
+    }
+
+    // Delete item by id
+    const onDeleteItem = (id) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.Item,
+            op: UndoOp.Delete,
+            oldList: subcollectionName,
+            newList: null,
+            data: itemsData
+        }]));
+
+        deleteDoc(doc(db, collectionPath + subcollectionName, id));
     }
 
     // Remove completed items
     const onRemoveCompleted = () => {
-        setData(data.filter((item) => !item.completed));
+        setUndoStack(undoStack.concat([{
+            type: UndoType.Item,
+            op: UndoOp.Delete,
+            oldList: subcollectionName,
+            newList: null,
+            data: itemsData
+        }]));
+
+        const batch = writeBatch(db);
```

Comment:
Why is it important that these are done in a single batch?

Reply:
I was thinking perhaps it's possible to get stuck in the middle of removing the completed items, so it's possible something like only half of them are removed.